src/main/scala/io/github/evis/scalafix/maven/plugin/mojo/ScalafixMojo.scala

```diff
@@ -84,19 +60,59 @@ final class ScalafixMojo extends AbstractMojo {
       getLog.info(
         "Skip scalafix since both skip.main and skip.test flags passed")
     } else {
+      val build = project.getBuild
+      val buildPath = getPath(build.getDirectory)
+      val mainSources = getSources(
+        skipMain,
+        "main",
+        buildPath,
+        mainSourceDirectories,
+        project.getCompileSourceRoots,
+        build.getSourceDirectory
+      )
+      val testSources = getSources(
+        skipTest,
+        "test",
+        buildPath,
+        testSourceDirectories,
+        project.getTestCompileSourceRoots,
+        build.getTestSourceDirectory
+      )
+      val mainOutputs =
+        if (skipMain) Seq.empty else Seq(build.getOutputDirectory)
+      val testOutputs =
+        if (skipTest) Seq.empty else Seq(build.getTestOutputDirectory)
       val params =
         List(
-          SourceDirectoryParam(sourceDirectory).ifNot(skipMain),
-          SourceDirectoryParam(testSourceDirectory).ifNot(skipTest),
-          ProjectDependenciesParam(projectDependencies),
-          CompiledDirectoryParam(compiledDirectory),
-          CompiledDirectoryParam(testCompiledDirectory),
-          PluginsParam(plugins),
+          SourceDirectoryParam(mainSources ++ testSources),
+          ProjectDependenciesParam(project.getArtifacts.asScala),
+          CompiledDirectoryParam(mainOutputs ++ testOutputs),
+          PluginsParam(build.getPlugins.asScala),
           ModeParam(mode),
           ConfigParam(config),
           CommandLineArgsParam(commandLineArgs)
         )
       run(params, getLog)
     }
   }
+
+  private def getSources(
+      flag: Boolean,
+      flagName: String,
+      outputDir: Path,
+      customDirs: JList[File],
+      projectDirs: JList[String],
+      projectDir: String): Iterable[File] = {
+    val dirs =
+      if (flag) Seq.empty
+      else if (!customDirs.isEmpty) customDirs.asScala
+      else
+        new File(projectDir, "/../scala") +: projectDirs.asScala.flatMap { x =>
```

Comment:
If `projectDirs` contain `../scala` directory too, the same `Processing ...` line gets printed twice. I suggest to get rid of duplicates before logging.

Reply:
@evis

- the logger is not visible outside of the mojo
- therefore, we cannot log in the `SourceDirectoryParam` code but in the mojo
- therefore, checking for existence, canonicalization and deduplication also must be in the mojo
- therefore, `SourceDirectoryParam` should not do any checking at all, and we need to remove some of its tests
would that work for you?

Reply:
I suggest extracting `getSources` to the separate module, but to leave logging inside mojo. This way, we can cover with tests the difficult logic of `getSources`.

Reply:
@kitbellew will you implement suggested changes to these tests?

If you're interested in merging this PR, but don't have time to tweak tests -- maybe it's better to leave the code as-is.

Reply:
i should finish it, i agree, sorry for the delay. give me a couple of weekends to look at it, please

Reply:
i have made the changes but need to figure out first the reason for https://github.com/SimonJPegg/mvn_scalafmt/issues/158, as this could be affected as well.

Reply:
ok, done. the problem in the other repo was due to unexpected behaviour of `new File(a, b)`.

Reply:
Thanks, I will review it during this week.

Reply:
i added an extra commit which applies the previous version of scalafix to itself, prompted by a build failure.

it seemed that since there's no `-SNAPSHOT` suffix in the pom, the build is generating the same version as the one recently published, and hence cannot find either (the new version is still being built, and the old one is masked).

perhaps #19 is related.